### ISSUE TITLE
Sanitizer error, example.com, report format 6, puppet version 4.8.2: required attribute not present: kind

### ISSUE THREAD

**Issue by larsen0815** [sodabrew/puppet-dashboard#395]
I recently switched to Debian Buster and upgraded puppet-dashboard from 2.0.0-beta2 to 3.0.0 afterwards. Reports from other hosts (Debian Stretch, Puppet 4.8.2) now fail with `Sanitizer error, example.com, report format 6, puppet version 4.8.2: required attribute not present: kind`.

I have compared the new installation directory to a backup and couldn't find any differences that might be causing this. Though, I noticed that "kind" is still present in the report on the client, but missing on the server:
````
# egrep "transaction_uuid|kind" $(puppet agent --configprint lastrunreport)
kind: apply
transaction_uuid: 2de17090-44f3-4824-8a74-2efd50c7aef1
````
````
# egrep "transaction_uuid|kind" /var/lib/puppet/reports/example.com/201910221521.yaml
transaction_uuid: 2de17090-44f3-4824-8a74-2efd50c7aef1
````

**Comment by ZeroPointEnergy** [comment]
Very strange. `kind` was removed from Puppet Report format 7 and should still be present in format 6. Is the Puppetserver doing some transformation on his own when sending the report to the dashboard?

**Comment by larsen0815** [comment]
Not that I know of...

**Comment by ZeroPointEnergy** [comment]
If you have a chance, can you please test if the following patch fixes the issue?

```
diff --git a/lib/puppet/report_sanitizer.rb b/lib/puppet/report_sanitizer.rb
index 8d824922..38ed1a20 100644
--- a/lib/puppet/report_sanitizer.rb
+++ b/lib/puppet/report_sanitizer.rb
@@ -359,6 +359,7 @@ module ReportSanitizer #:nodoc:
     def sanitize(raw)
       sanitized = super
       raw['master_used'] ||= nil
+      raw['kind'] ||= 'apply'
       Util.verify_attributes(raw, %w[noop noop_pending corrective_change master_used])
       Util.copy_attributes(sanitized, raw, %w[noop noop_pending corrective_change master_used])
     end
```

**Comment by larsen0815** [comment]
Patched, but I still get the same error:
````
Importing report at 2019-10-24 14:42 CEST
Sanitizer error, tamoto.atlantismedia.de, report format 6, puppet version 4.8.2: required attribute not present: kind
Backtrace
lib/puppet/report_sanitizer.rb:98:in `block in verify_attributes'
lib/puppet/report_sanitizer.rb:97:in `each'
lib/puppet/report_sanitizer.rb:97:in `verify_attributes'
lib/puppet/report_sanitizer.rb:249:in `sanitize'
lib/puppet/report_sanitizer.rb:277:in `sanitize'
lib/puppet/report_sanitizer.rb:294:in `sanitize'
lib/puppet/report_sanitizer.rb:343:in `sanitize'
lib/puppet/report_sanitizer.rb:360:in `sanitize'
lib/puppet/report_sanitizer.rb:23:in `sanitize'
app/models/report.rb:119:in `create_from_yaml'
````

Would it be helpful if I sent you such a report as a file?

**Comment by ZeroPointEnergy** [comment]
Yes that would be great

**Comment by larsen0815** [comment]
Not completely sure if there is anything confidential in those reports after having replaced some obvious data, so the password for the zip file is at https://pastebin.com/wrCp8xwx for 24 hours.

[1temp.zip](https://github.com/sodabrew/puppet-dashboard/files/3767989/1temp.zip)

**Comment by larsen0815** [comment]
@ZeroPointEnergy : Do you expect a patch to be released soon? Otherwise, I would switch my clients to the puppetlabs-version to workaround that problem.

**Comment by ZeroPointEnergy** [comment]
Sorry, saw your post with the reports zip too late. I don't really have time to work on a fix at the moment. I can point you to the code where the problem occures though, maybe you can figure something out:
- https://github.com/sodabrew/puppet-dashboard/blob/9310c5c898d8a02414f74801f3678a2a8b9de22e/lib/puppet/report_sanitizer.rb#L295 this is where the attribute gets added to the list of sanitized attributes.
- https://github.com/sodabrew/puppet-dashboard/blob/9310c5c898d8a02414f74801f3678a2a8b9de22e/lib/puppet/report_sanitizer.rb#L389 here we add it because it was removed in format 7, maybe that has to be in format 6 in your case in the class above this one.

**Comment by larsen0815** [comment]
I had no luck trying things. Will switch my older clients to the puppetlabs version instead of using the distribution's one.

**Comment by larsen0815** [comment]
TL;DR
Found a workaround

1. The puppetlabs version for Debian Stretch is only 4.10, so still at report format 6.
2. The previous tests I made would have been sufficient, had I noticed that I not only have to restart puppet-dashboard and puppet-dashboard-workers, but also kill the worker process that apparently don't get stopped correctly. As a result of this, my code changes didn't come into effect. I only noticed this after I tail'ed the log (that I now have found), and added some debugging messages (that didn't appear until I found out about the restart problem)

Your patch from above almost solves the problem, but it has to be placed before the line "sanitized = super".

**Comment by larsen0815** [comment]
FYI: Differences between report files on client and master.

Client, at the end:

````
host: example.com
time: 2019-11-05 11:40:42.568701611 +01:00
kind: apply
report_format: 6
puppet_version: 4.8.2
configuration_version: 1572885531
transaction_uuid: 7da01705-da4f-4133-a675-74b04c6d029a
code_id:
catalog_uuid: 5d284e55-0b91-42f0-b16c-bbf76efc7a7a
cached_catalog_status: not_used
master_used:
environment: production
status: unchanged
noop: false
noop_pending: false
corrective_change: false
````


Master, at the top:
````
host: example.com
time: '2019-11-05T11:40:42.568701611+01:00'
configuration_version: 1572885531
transaction_uuid: 7da01705-da4f-4133-a675-74b04c6d029a
report_format: 6
puppet_version: 4.8.2
status: unchanged
transaction_completed:
noop: false
noop_pending: false
environment: production
````

Master, at the end:
````
corrective_change: false
catalog_uuid: 5d284e55-0b91-42f0-b16c-bbf76efc7a7a
cached_catalog_status: not_used
````

Seems like puppet-master is changing those files. Though, I don't know if this already happened with the previous version with Debian Stretch.

**Comment by ZeroPointEnergy** [comment]
Should be fixed in 3.0.1
